scipy/spatial/_alpha_shapes.py

```diff
@@ -0,0 +1,193 @@
+"""
+Alpha Shapes Code
+
+.. versionadded:: 1.6.0
+
+"""
+import itertools
+import numpy as np
+from scipy.spatial import Delaunay
+from scipy._lib._disjoint_set import DisjointSet
+
+
+__all__ = ['AlphaShapes']
+
+
+def _calculate_circumspheres(points, simplices, rcond):
+    # circumcenter, c, satisfies:
+    # |p_i - c| = |p_j - c| for all i and j in [1 .. ndim+1]
+    # |p_i|^2 - 2.p_i.c + |c|^2 = |p_j|^2 - 2.p_j.c + |c|^2
+    # |p_i|^2 - 2.p_i.c = |p_j|^2 - 2.p_j.c
+    # |p_i|^2 - |p_j|^2 = 2.(p_i - p_j).c
+
+    # build overdetermined system of linear equations
+    tetrahedra = points[simplices]
+    A = 2 * (tetrahedra - np.roll(tetrahedra, 1, axis=1))
+
+    squared_norms = np.einsum('ij,ij->i', points, points)[simplices]
+    b = squared_norms - np.roll(squared_norms, 1, axis=1)
+
+    # solve stack of linear systems of equations with `pinv` (`rcond` handles
+    # rank-deficient simplices).
+    Ainv = np.linalg.pinv(A, rcond=rcond)
```

Comment:
@ilayn I'm using `pinv` to solve a stack of least-squares problems. Is there a better way? I would use `lstsq` but it doesn't support stacks and looping is too slow.

Reply:
For now, I guess we have to live with it. `pinv` uses SVD so I am surprised that loop is slower. But it is what it is I guess.